### PR TITLE
Mesh component

### PR DESCRIPTION
**Pull request by Tjalling-dejong**
## Issue addressed
Adresses issue #671 
Implementation of the ModelComponent class for the mesh component.

## Explanation
Explain how you addressed the bug/feature request, what choices you made and why.

## Checklist
- [x] Updated tests or added new tests
- [ ] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [ ] Updated documentation if needed
- [ ] Updated changelog.rst if needed
- [ ] For predefined catalogs: update the catalog version in the file itself, the references in data/predefined_catalogs.yml, and the changelog in data/changelog.rst

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

**Comment by Tjalling-dejong** [comment]
@hboisgon I am encountering an issue with the get_mesh() method, I saw through git blame that you made the last commit to this code. I have not changed the code. Now that we are using XUgrid version 0.0.9 the code breaks. I did not manage to figure it out if the bug is from our side or in XUgrid. Could you maybe have a look?

**Comment by hboisgon** [comment]
> @hboisgon I am encountering an issue with the get_mesh() method, I saw through git blame that you made the last commit to this code. I have not changed the code. Now that we are using XUgrid version 0.0.9 the code breaks. I did not manage to figure it out if the bug is from our side or in XUgrid. Could you maybe have a look?

I see we get the same error in the hydromt_delft3dfm test. I'll have a look but in main and not v1

File: hydromt/models/__init__.py
Comment (deltamarnix) [inline hydromt/models/__init__.py:12]: Remove?

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:42]:
```suggestion
    def set(
        self,
        data: Union[xu.UgridDataArray, xu.UgridDataset],
        *,
        name: Optional[str] = None,
        grid_name: Optional[str] = None,
        overwrite_grid: bool = False,
    ) -> None:
```

I would make all naming mandatory

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:89]: Is it still needed? ;)

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:69]: What would that entail? I can imagine that we want one grid per component.

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:75]: Remove?

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:61]: I expected the first call to be `_initialize_mesh()`, just like other components.

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:116]:
```suggestion
    self._model_root._assert_write_mode()
```

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:149]:
```suggestion
    self._model_root._assert_read_mode()
```

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:152]:
```suggestion
                fn, root=self._model_root.path, logger=self._logger, **kwargs
```

Do you have tests for these, or should you update from the refactor-models branch?

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:171]: In my branch `region-component-2`, I am trying to get rid of the `region: dict`. I think this will influence this piece of code as well.

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:286]: Shouldn't this be handled by the `region` component instead?

File: hydromt/models/components/mesh.py
Comment (deltamarnix) [inline hydromt/models/components/mesh.py:394]:
```suggestion
    def add_data_from_rasterdataset(
        self,
        *,
        raster_fn: Union[str, Path, xr.DataArray, xr.Dataset],
        grid_name: Optional[str] = "mesh2d",
        variables: Optional[list] = None,
        fill_method: Optional[str] = None,
```

File: tests/conftest.py
Comment (deltamarnix) [inline tests/conftest.py:373]: There are already some other mocks for a Model. Not sure if you want to put this in conftest, as it will be different per test case.

File: tests/conftest.py
Comment (deltamarnix) [inline tests/conftest.py:379]: Let's mock the other parts as well

File: tests/conftest.py
Comment (deltamarnix) [inline tests/conftest.py:373]: Try using `def mock_model(tmpdir, mocker: MockerFixture):`, so that you can use the `pytest-mock` wrapper for making your mocks.

**Review by deltamarnix — COMMENTED** [review]
This is my initial review. I will probably come with more in a second round, but that also depends on my own experiences with creating components.

Let's have another look at how we mock in our repository. I prefer using `pytest-mock`, where we make use of `MockerFixture`. In that case we don't need to write things like: `with patch(...):`.

File: hydromt/models/plugins.py
Comment (savente93) [inline hydromt/models/plugins.py:17]: I can see this code is not up to date with `model-refactor`. if we're going to wait to get that in before you merge this you might want to merge it because I can already see it's going to cause a few merge issues here.

File: hydromt/models/components/mesh.py
Comment (savente93) [inline hydromt/models/components/mesh.py:591]: Fix todo, or make issue?

**Review by savente93 — COMMENTED** [review]
Nice tests. I have a few small comments but once it's up to date it should be very close.

File: tests/conftest.py
Comment (Tjalling-dejong) [reply tests/conftest.py:373]: I'm using this fixture in test_grid_component as well. Should I use I put this fixture in both test_mesh_component.py and test_grid_component.py?

File: hydromt/models/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/models/components/mesh.py:69]: I don't know. The TODO was already there in the original code and left it in for now.

File: hydromt/models/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/models/components/mesh.py:61]: I think it is done this way because you cannot create an empty UgridDataset

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:18]: I think you still need the __all__ but his time with MeshModelComponent.
@savente93 previously xugrid was an optional package. How does this work now with components discovery? Can we still make this one discovered only if xugrid is installed? Or should in the end xugrid become a mandatory dependency of hydromt ?

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:81]: Nice that you did a new function, I would however also include updating crs in that function as well instead of here.

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:83]: You'll need to add ``@hydromt_step`` decorator to all methods available for build and update ie read/write/create/add_data

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:162]: This is to create a 2D mesh so I would keep it in the function name. the current MeshComponent handles multiple grids which can be 1D or 2D or 3D.

```suggestion
    def create2d(
```

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:370]:
```suggestion
    @hydromt_step
    def add_2ddata_from_rasterdataset(
```

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:372]: Maybe the * can go after raster_fn? We're already not happy with the name :)

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:377]: It can still be a list right? In case method is different per variable in ``variables``

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:446]:
```suggestion
    @hydromt_step
    def add_2ddata_from_raster_reclass(
```

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:566]:
```suggestion
                raise ValueError("Data and Mesh should have the same CRS.")
```

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:268]: Since you removed ``region`` maybe you could create an equivalent ``mesh_region`` property? 
```suggestion
    @property 
    def mesh_region(self) -> Optional[gpd.GeoDataFrame]:
        """Return mesh total_bounds as a geodataframe."""
        if self.data is not None:
            region = gpd.GeoDataFrame(
                geometry=[box(*self.data.ugrid.total_bounds)], crs=self.crs
            )
            return region
        return None
```

**Review by hboisgon — CHANGES_REQUESTED** [review]
Nice work @Tjalling-dejong especially with the extra testing!
A few comments to address before merging but the refactor PR changed a little again so I think you'll have to wait for this one to be merged first before you can address everything.

File: hydromt/models/components/mesh.py
Comment (hboisgon) [reply hydromt/models/components/mesh.py]: No but then I suggested to add a new ``mesh_region`` property to replace the old ``region`` one. It will be quite useful to update ``MeshComponentRegion`` as we discussed.

File: hydromt/models/components/mesh.py
Comment (hboisgon) [reply hydromt/models/components/mesh.py:69]: For now ``MeshComponent`` is still multi-grid. We can see later if we want to add single grid Component like ``Mesh1dComponent``, ``Mesh2dComponent`` and ``Mesh3dComponent``. The TODO can be removed and open as an issue if anyone complains. It's just that for the set methods I only allow to update data or grid schematization one grid at a time.

File: hydromt/models/components/mesh.py
Comment (hboisgon) [reply hydromt/models/components/mesh.py:61]: Indeed! So the choice would either be xr.Dataset or None but I think None is easier and clearer to handle.

File: hydromt/models/components/mesh.py
Comment (hboisgon) [reply hydromt/models/components/mesh.py]: Why was the is_overwrite_mode added? Calling the whole ``set`` method means you are overwriting anyway if not the grid then the data...

File: hydromt/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/components/mesh.py:19]: XUgrid will become a mandatory dependency in v1

File: hydromt/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/components/mesh.py:376]: I based the typing of the typing in mesh2d_from_rasterdataset. There it is Optional[str], but I see in the docstring now that a list is also possible. I'll correct it in mesh2d_from_rasterdataset as well

File: hydromt/components/mesh.py
Comment (deltamarnix) [inline hydromt/components/mesh.py:26]: I believe that this `logger` can go

File: hydromt/models/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/models/components/mesh.py]: I removed is_overwrite_mode. Not sure why I added that

File: hydromt/components/mesh.py
Comment (deltamarnix) [inline hydromt/components/mesh.py:86]: I would opt for a `DEFAULT_FN = "mesh/mesh.nc` at the top of this file and use it in `fn` for both the `read()` and `write()` functions.

File: hydromt/components/mesh.py
Comment (deltamarnix) [inline hydromt/components/mesh.py:238]: I miss this `_initialize_mesh` function that we also see in the grid component. Would be nice to have it consistent.

File: hydromt/components/mesh.py
Comment (deltamarnix) [inline hydromt/components/mesh.py:256]: I would expect this in a setter, not in a getter.

File: hydromt/components/mesh.py
Comment (deltamarnix) [inline hydromt/components/mesh.py:282]: Would be nice if your editor did autoformatting. We have enabled that in vs code settings. Perhaps you have some overwritten settings in your vs code personal settings. I expect some new lines in between

File: hydromt/components/mesh.py
Comment (deltamarnix) [reply hydromt/components/mesh.py:379]: It almost looks like these are two different functions. One for a file, one for an actual dataset

File: hydromt/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/components/mesh.py:282]: Weird that it did not autoformat this time

File: hydromt/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/components/mesh.py:261]: I think this feature can go. There is a check in _add_mesh that checks if the crs is uniform.

**Comment by Tjalling-dejong** [comment]
@deltamarnix @hboisgon @savente93  Thanks for the feedback!
I added your suggestions and updated code + tests. 
There are still multiple breaking tests but these are related to plugins and circular imports. I will look into that tomorrow.

File: docs/dev/migrating-to-v1.rst
Comment (hboisgon) [inline docs/dev/migrating-to-v1.rst:243]: Table format looks off. Can you check? All columns length should match ie "|" should be aligned. If it was because of my suggestion then it's because I cant easily check alignment in this github view sorry..

File: docs/dev/migrating-to-v1.rst
Comment (hboisgon) [inline docs/dev/migrating-to-v1.rst:367]: Same here for table alignement

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:39]: We just discussed yesterday that we want to pass a default filename in the component __init__ and then use None in the argument for read and write. Can you update?
You can check out the TablesComponent PR

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:241]:
```suggestion
            self._initialize()
```

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:247]: Does this work now? Then nice we can initialize even with an empty grid :)
You'll need to add an initialize step in the read function. You can check maybe TablesComponent to see how initialize is used

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:254]: Maybe you need to update all the properties now no? Because with initialize you will get an empty xugrid object and not None anymore for data. Check this also works for empty xugrid object.
Please update all properties like bounds etc accordingly

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:267]: I think this will be named region again in @deltamarnix PR on regions so maybe you could change the name already or check with him how to best update?

File: docs/api.rst
Comment (hboisgon) [inline docs/api.rst:595]:
```suggestion
   MeshComponent.read
   MeshComponent.get_mesh
```

File: hydromt/components/mesh.py
Comment (hboisgon) [inline hydromt/components/mesh.py:378]: I still miss in the name that this method works for 2D grids and not 1D and 3D. I previously suggested ``add_2ddata_from_rasterdataset``. I'm never good with names but I still think we should highlight this fact somehow in the name.

**Review by hboisgon — CHANGES_REQUESTED** [review]
Almost there @Tjalling-dejong ! Nice that you could add an initialize function for mesh! But I think you still need to update a couple of things to fully take it into account.
I wonder if we are not missing model update tests in our tests benches? Ie model opened in r+ mode and check if the existing mesh/grid/other components are indeed read properly before being updated. If these tests are in place please ignore.

File: hydromt/components/mesh.py
Comment (Tjalling-dejong) [reply hydromt/components/mesh.py]: I renamed it to `add_2d_data_from_rasterdataset`

File: tests/components/test_mesh_component.py
Comment (hboisgon) [reply tests/components/test_mesh_component.py:206]:
```suggestion
    component._data = None
    component.read()
    assert component.data == data
    
    m2 = Model(root=str(tmpdir), mode="r+")
    # Sample data
    data = xu.data.elevation_nl().to_dataset()
    data.grid.crs = 28992
    data.name = "elevation_v2"
    m.mesh.set(data=data, grid_name="elevation_mesh")
    
    # check both data variables are still there ie mesh was read before being updated
    assert "elevation_v2" in m2.mesh.data
    assert "elevation" in m2.mesh.data # I think this is the correct name?
    
```

File: docs/dev/migrating-to-v1.rst
Comment (hboisgon) [inline docs/dev/migrating-to-v1.rst:379]: Alignment still looks off? Or is it just this view?

**Review by hboisgon — CHANGES_REQUESTED** [review]
Thanks for the changes @Tjalling-dejong. Almost there but still a couple of steps to add to fully take advantage of the new ``_initialize`` function.
I also explained the test I wanted in a little more details, sorry it was not clear enough.

If you update the initialize part then we are good to merge :)

File: docs/dev/migrating-to-v1.rst
Comment (Tjalling-dejong) [reply docs/dev/migrating-to-v1.rst]: It is just the view, if you check this file in the branch the table looks fine

**Review by hboisgon — APPROVED** [review]
Looking all good now, thanks @Tjalling-dejong :)

**Pull request merged**
